core/unix/rseq_linux.c

```diff
@@ -646,41 +675,31 @@ rseq_locate_tls_offset(void)
+         * app library segment base on x86, and negative on aarchxx. However, in the
+         * absence of rseq support from glibc, the app manually registers its own
+         * struct rseq which is present in static TLS, which is at a negative offset
+         * from the app library segment base on x86, and positive on aarchxx.
          */
-        for (i = 0; IF_X86_ELSE(addr - i * alignment >= seg_bottom,
-                                addr + i * alignment < seg_bottom + seg_size);
-             i++) {
-            byte *try_addr = IF_X86_ELSE(addr - i * alignment, addr + i * alignment);
-            ASSERT(try_addr >= seg_bottom); /* For loop guarantees this. */
-            /* Our strategy is to check all of the aligned static TLS addresses to
-             * find the registered one.  Our caller is not supposed to call here
-             * until the app has registered the current thread.
-             */
-            static const int RSEQ_RARE_SIGNATURE = 42;
-            int res = dynamorio_syscall(SYS_rseq, 4, try_addr, sizeof(struct rseq),
-                                        RSEQ_FLAG_UNREGISTER, RSEQ_RARE_SIGNATURE);
-            LOG(GLOBAL, LOG_LOADER, 3, "Tried rseq @ " PFX " => %d\n", try_addr, res);
-            if (res == -EINVAL) /* Our struct != registered struct. */
-                continue;
-            /* We expect -EPERM on a signature mismatch.  On the small chance the app
-             * actually used 42 for its signature we'll have to re-register it.
+        ASSERT(seg_bottom <= addr && addr < seg_bottom + seg_size);
+        for (i = (seg_bottom - addr) / alignment;
+             addr + i * alignment < seg_bottom + seg_size; ++i) {
+            byte *try_addr = addr + i * alignment;
+            ASSERT(seg_bottom <= try_addr &&
+                   try_addr < seg_bottom + seg_size); /* For loop guarantees this. */
+            /* Our strategy is to check all of the aligned addresses to find the
+             * registered one.  Our caller is not supposed to call here until the app
+             * has registered the current thread (either manually or using glibc).
              */
-            if (res == 0) {
-                int res = dynamorio_syscall(SYS_rseq, 4, try_addr, sizeof(struct rseq), 0,
-                                            RSEQ_RARE_SIGNATURE);
-                ASSERT(res == 0);
-                res = -EPERM;
-            }
-            if (res == -EPERM) {
-                /* Found it! */
+            if (try_struct_rseq(try_addr)) {
                 LOG(GLOBAL, LOG_LOADER, 2,
                     "Found struct rseq @ " PFX " for thread => %s:%s0x%x\n", try_addr,
-                    get_register_name(LIB_SEG_TLS), IF_X86_ELSE("-", ""), i * alignment);
-                offset = IF_X86_ELSE(-i * alignment, i * alignment);
+                    get_register_name(LIB_SEG_TLS), (i < 0 ? "-" : ""),
+                    abs(i) * alignment);
```

Comment:
Where is this `abs` coming from?  If it resolves to a libc call that is not going to work.  Are you assuming it is a macro?  It is probably safest to define our own macro and avoid some platform having it as a libc import.

Ditto for the use below.

Reply:
Done.